### DISCUSSION TITLE
No type-safety for selector's projector function

### DISCUSSION THREAD

**Discussion by dzonatan** [ngrx/platform#3097] [Ideas]
<img width="735" alt="image" src="https://user-images.githubusercontent.com/5166666/72372715-83977600-370f-11ea-9d66-c3c00e51ddf6.png">

Is there a reason why projector function does not have a type-safety for its parameters yet?
This quickly becomes a nightmare when having a lot of unit tests for complicated projectors.
I've looked into internals and found that for some reason the [`MemoizedSelector`](https://github.com/ngrx/platform/blob/2cc88857421b74e01b816f99fe28234bbe77ff1f/modules/store/src/selector.ts#L34) is using a default type [`DefaultProjectorFn`](https://github.com/ngrx/platform/blob/2cc88857421b74e01b816f99fe28234bbe77ff1f/modules/store/src/selector.ts#L17) for its projector which does not know anything about parameter types but only `Result`:

```ts
export function createSelector<State, S1, Result>(
  s1: Selector<State, S1>,
  projector: (s1: S1) => Result
): MemoizedSelector<State, Result>; // <-- third type is a `DefaultProjectorFn`
```

This could be easily fixed by explicitly providing a type of projector function (for every `createSelector` of course):
```ts
export function createSelector<State, S1, Result>(
  s1: Selector<State, S1>,
  projector: (s1: S1) => Result
): MemoizedSelector<State, Result, (s1: S1) => Result>;
```

<img width="692" alt="image" src="https://user-images.githubusercontent.com/5166666/72373608-4af89c00-3711-11ea-8e58-df1fcfcd5188.png">

Maybe this was already discussed? Are there any hidden issues that I'm not aware of? 
Otherwise it's strange why this was not escalated yet.

## If accepted, I would be willing to submit a PR for this feature

[x] Yes
[ ] No

I have all the changes ready to be reviewed - just let me know if the solution here is appropriate.

**Comment by timdeschryver** [comment]
The change looks good to me!

**Comment by alex-okrushko** [comment]
So even the usage of `DefaultProjectorFn` I had to revert within Google's repo to `AnyFn`, because  some of the quirks of the TS. For example, in TS accessing Array by Index always gives you the result with the type and never `undefined`, which is clearly not what it should be (microsoft/TypeScript#13778)

I tried to demonstrate [the problem here](https://www.typescriptlang.org/play/index.html#code/PTAEFpK6dv4YgUCUheDcOI7oC2BLAdrtgIYA2oAcgOYBKAHqAM4CmpzAxgC4D2ATqEQAObbM3ydinXN3xJmdQX06hOAT0HNQAZVYcevADwAVADSgAagD5QAXlAAKRhM7MAXKCMBKW9fMBuOQUlFXVNABFmADNiAFdSTgAFXm4AKz0+ADF8Y2s7ewA6QuJeKkZ3YnxVAG0AXW8bayMAwMVeZQIXXmj2TQBZZmxuXAAvZgATHTYuPgMkUG1nZhM50BpmRjjOZfmk1PTeLNtQCOjN3bTpg+y1jfjLJGt5F3wxxm1dS4MtRbMbzesAN4rXi6YgseyedwANyGYwC80EyQu+ncBSKJTKoAq1TqPlW6028KYzE4f3iqJBt04AH53GTOPVrDDcHCVuw2MV6ajGaBmayAL5IFrBSIxfBcaT4UDsEGSZiTfZfH7aACMvwJd3sK3mjBV7gVn2+crMWhVlm281AiL2l1Ruv1Kp59KQkNA-UGI3GBv0SuN+KpgNA2ulMic0rIpGIACM2EdHPrFjzrci+I4VY5PJ4ifMQZwYrwpQB5KPI-Jgxi4Kj4ezsCPRthmIGW5tWpH7C0t0Agjksbl4gGgfkdlssUkazgU8e0-2bHkDofB5vs5ic8d9hqgefD-lZpCC1BoQBiOzgCEQyJRaAwWFN9AJsMIBmJnJKUGBEO+P5+EEKOswusQelAYsUk3INmxZdwnF4AgqACfcwB4QRQDYKFWGJG8+FASJMM4AALTQAEFeF4YhVFAbhIiAgAhAApABRABhIwtCQdhQ2Ua89CIkiyLyV1gNqPEqgHCDQAAcmIMTB0bAQxncMSoykodQNEsT2CUmpmlY9isS4GIyG9TC7BlFcXEM3gtWbTiuG40jh1QTtLUwUB8G4KRALwyRQAAd00ACekYN48M0AgxnkFRcK87zuDiMYuxJfMpTFMLIgIcZg2KfgN0yqoVQABjympQAAH2Kly4lIYNUAACW4XzUP4AgPC0XSAorfAqCxYjSNAKMyNC8LvNwUhyDIbzSLeKhcFQ0BVBiiLNEpTYqrAIa8K6niQg0LEXhc5gGvE5KojSsYxLMRhuGaiKCAAazeAADUrypG+6fOG8h8H2v9QEiwQNFkTwhQcpysGCgR8EiP8QTitRttwN5gJ2uLXOUYDivwCrtPwMMlviI4ALzAyPn0fJk32exhNk+TJMHOpmlQAipTw+G3pG3rNDB60YwGcHQBosE8E+lR1k4N4kDC9lik0UVxSkGRQHkDQuByewJw8V17AHHg6IAR300gKXcLw8SjbhuA5KUd3psBHNtxz8JBBbW24bnsAEN5cMrXDSC9lw4swFo9HsAn9fM0m20uCmRLk8Saf5OpPHybW9bIewMZG3dUDo4i+HcIiqBiURxHIyjYc0MT09IKSWZRrFAsrfB6w5y7BGKYhRE6EutvL4CxKFIA)

Of course, there is a solution - to give the explicit return type to the projection function, and maybe that's the right solution, but it's hard when this change is introduce later into a huge codebase.

What I what to highlight here is that `.projection(...)` is used in tests only and if you pass something that is mistyped into it - you'll most likely get an error. Yes, it'll be a runtime error of the failing test and not compile time error, but "fixing" it might create more problems upgrading to a newer version of NgRx. Is it worth it? I'm not sure.

**Comment by dzonatan** [comment]
I hear you and agree that this is kinda a breaking-change. Even the example-app tests in this repo has a few usages of projector func that would break. But by not having type-safety here we're losing other perks like language-service support - rename and other stuff.. now when refactoring you have to do it in old way with find & replace. As you said - most of the time this will result in a runtime error, but not every - so because of these few times you still have to go through all your usages and fix it manually, otherwise your codebase soon will lose integrity.

Also `projector` function is not technically limited to tests only. I've seen some cases when it was used in the implementation, although I'm not convinced it is a good practice.

**Comment by dev054** [comment]
Hey, any chance to revisit it before v11? It's a pain to use projector without type-safety.

**Comment by NicBright** [comment]
I've run into the same typings problem and wished the `.projector` was properly typed.

Quoting @alex-okrushko 
> What I what to highlight here is that .projection(...) is used in tests only and if you pass something that is mistyped into it - you'll most likely get an error. Yes, it'll be a runtime error of the failing test and not compile time error, but "fixing" it might create more problems upgrading to a newer version of NgRx. Is it worth it? I'm not sure.

Yes, fixing this issue is definitely worth it, because NgRx itself is a project written in TS. For my part, this makes me expect that types should be handled properly, even in tests.

Regarding the breaking change dilemma. When you don't want to break existing functions, there's a classic solution: provide the fixed version under a new name, deprecate the old name, wait for a given period and then remove the old version.

You can even do it the other way round (with a little bit of extra effort): rename the current version to maybe `createSelectorDeprecated` (and provide the fixed version under the current name `createSelector`). Additionally, provide a script  that automatically updates all imports with an alias (I'm thinking of an update helper similar to `ng update`).

Previously: `import { createSelector } from '@ngrx/store'`
With alias: `import { createSelectorDeprecated as createSelector } from '@ngrx/store'`

**Comment by david-shortman** [comment]
What about introducing a new test method instead that is type safe @alex-okrushko?

proposed tiny change: https://github.com/david-shortman/platform/commit/e7065b182c243cd365428df85e8f758ee048ab52

By introducing a (hopefully someone could come up with a better named) function like `testProjector`, we could be able to do the following without any need to break and migrate existing code, and it would be strongly typed:

```typescript
const selectSomeSum = createSelector(a => a, b => b, (a, b) => a + b);
const result = testProjector(selectSomeSum, 1, 2); // 3
```

this would be enabled by the createSelector updates in #3023 which make inferring the results of the selectors of which a selector is composed easy.

**Comment by MaximSagan** [comment]
Just created [an issue that was closed as a duplicate of this](https://github.com/ngrx/platform/issues/3196). Let's not sit on this for another two years. If #3023 (which I really hope will be included in v13) already introduces breaking changes in the `MemoizedSelector` type, it seems like a good opportunity to include this too.

> Of course, there is a solution - to give the explicit return type to the projection function, and maybe that's the right solution, but it's hard when this change is introduce later into a huge codebase.
>
> What I what to highlight here is that .projection(...) is used in tests only and if you pass something that is mistyped into it - you'll most likely get an error. Yes, it'll be a runtime error of the failing test and not compile time error, but "fixing" it might create more problems upgrading to a newer version of NgRx. Is it worth it? I'm not sure.

+1 Definitely worth it. I don't even see it being a cost for a huge codebase, even if they've been using `.projector()` extensively in their unit tests. They're getting something that they lacked before, and something that quite frankly they should be expecting in a Typescript project: type safety.

**Comment by david-shortman** [comment]
We can make this a gentle transition.

The `projector` method can have two signatures. One which is strictly typed, and one which is loosely typed.

Then when this change is introduced in some version, it is not breaking, but the loosely typed signature can be marked as deprecated.

Then in the subsequent version, the deprecated signature is removed.

I'd be happy to open a PR with this approach.

**Reply by david-shortman** [reply]
yolo, went for it: https://github.com/ngrx/platform/pull/3223

**Comment by david-shortman** [comment]
I've authored a PR that demonstrates a soft introduction for strictness that doesn't affect production code or existing tests.

https://github.com/ngrx/platform/pull/3360

![](https://user-images.githubusercontent.com/25187726/160259614-85dbda9c-b520-4c32-847f-51e344290b3f.png)

**Comment by david-shortman** [comment]
@timdeschryver It looks like v14 is approaching with many new PRs. Is there interest in tackling this in v14? My currently open PR uses an opt-in strategy and suggests a path for future migration to carefully take people toward strict projectors.

**Reply by MaximSagan** [reply]
There is interest!
@brandonroberts Let's make it happen!

**Reply by brandonroberts** [reply]
We'll make it happen if it's the right choice for the majority of users and there's a way to keep using the previous style if necessary.

I think we can come up with something better than 'strict' for the types as it doesn't seem like something I've seen in other projects for types.

**Reply by david-shortman** [reply]
> I think we can come up with something better than 'strict' for the types as it doesn't seem like something I've seen in other projects for types.

I was aiming to make it "opt-in", but we could follow the pattern Angular is using for introducing strongly types forms.

Projectors could become strongly typed, and one could opt out by using `selector.projector<any>`. We could provide a schematic that automatically migrates existing projectors to this usage. Then developers could make their own choice about migration from there.

**Reply by david-shortman** [reply]
> I think we can come up with something better than 'strict' for the types as it doesn't seem like something I've seen in other projects for types.

Just spitballing another idea here, but we could also introduce strong types as an alpha version or another tagged version that is published alongside regular updates.

Then there would be no use of a generic parameter with `projector`, but one could start using strict projectors by using a version like 14.0.0-strict, for instance.

**Reply by david-shortman** [reply]
> I think we can come up with something better than 'strict' for the types as it doesn't seem like something I've seen in other projects for types.

A final alternative could be adding an eslint rule. While it wouldn't break compilation, it would allow developers to start getting warnings for projectors that are incorrectly used (or errors if they want to be stricter).

**Reply by brandonroberts** [reply]
I'd be in favor of the lint rule as a warning as a gentle introduction for the strict projector. The projector as it is today has been battle tested by Google which I consider to be the most strict consumer of the library.  They have not expressed interest publicly anyway about making the projector more strict.

**Reply by NicBright** [reply]
> I'd be in favor of the lint rule as a warning as a gentle introduction for the strict projector.

Please let me argue against this suggestion:

- more than two years passed since this thread was opened 
- making a "gentle introduction" sounds to me that we can expect this feature to need another two years until it can finally land
- using a lint rule for something that can be clearly and better pin-pointed directly by the ts compiler sounds like an antipattern to me

Here's an easy way to introduce it in a non-breaking way:
- add another method to the selector, let's call it `test` (or `run` or `execute` or whatever)
- make this method be strict, and leave the `projector` just as is
- change the docs and promote the new `test` method for type-safe testing selectors

**Reply by timdeschryver** [reply]
To give an update, we're trying to find out how breaking this change would be before we take further actions.

**Reply by david-shortman** [reply]
I'm up for the challenge of creating a type-based eslint rule if that is considered to be the smoothest transition path. Feel free to ping if a direction is chosen.

**Reply by aokrushk** [reply]
Alright, I'm now in the camp of strongly supporting the proper typing of `projector` function.

We also ran into this issue and tests didn't catch the mistyped property.

**Reply by david-shortman** [reply]
Personally, I think the simplest route would somewhat mirror Angular's strictly typed forms.

1. Refactor createSelector to strongly type the selector's `projector`
2. Migrate existing codebases with a schematic to change calls to use loose typing, like `someSelector.projector<any>(...)`
3. Developers can remove the generic argument `any` and the projector will be strongly typed

**Reply by timdeschryver** [reply]
Thanks for the idea @david-shortman, I like it!
We're verifying the last pointers, but I think we should try to land this in v15.

**Reply by timdeschryver** [reply]
@david-shortman and others, we discussed this and we want to go forward with these changes.
If possible, these would be included in the v15 release.

I created two issues for this:
- Typed projectors: https://github.com/ngrx/platform/issues/3571
- Migration: https://github.com/ngrx/platform/issues/3572

Thanks for the input!
